Merging Boards & Code Cleanup

juanfxv opened the pull request:
- Changed Background image
- Added `plant.png` image
- Added comments
- Implemented plant functionality
- Some code cleanup

icourt committed:
Unmerge files, unzip code.

icourt committed:
Clean up main branch

juanfxv committed:
Implemented plant functionality merged boards together now needs collision.

icourt committed:
Merge branch 'main' into matthew

icourt committed:
Merge

icourt committed:
Refactor code structure, introduce type annotations, and improve readability

- Organized project structure with dedicated modules for entities and managers.
- Added type annotations for variables and function parameters.
- Centralized image loading and scaling in a utility function.
- Encapsulated entity creation and initialization logic.
- Created a GameController for accessing game states from entity classes
- Implemented a fade-to-black effect when the game is over.
- Fixed bugs with allowing multiple plants on the same square & zombies being able to "skip" some plants

icourt committed:
Rename gridtest -> main file

icourt commented:
1. **Modular Code:**
   - Moved entity classes (Plant, Zombie, SpeedyZombie) and their related constants to a separate module (`entities/`).
   - Moved game management logic to `GameController` class in a separate module (`managers/`).
   - Moved all images to `assets/images/` and sounds to `assets/sounds/`
   - Renamed `gridtest.py` to `main.py`

2. **Better Image Loading:**
   - Created a function `load_and_scale_entity_images` to handle loading and scaling of entity images automatically based on the module
   - Created a function `draw_grid_and_entities` to draw background, grid, and all entities at one time. Also ensures Zombies appear on top of Plants

3. **Game Initialization:**
   - Plant + Zombie collision functionality via @juanfxv 
   - Moved pre-defined zombies list to a `initialize_zombies` function to remove duplicate code.
   - Implemented a "GameController" Class to allow modifying the control lists (`zombies`/`plants`) from any class, see `entities/zombie.py` for example.
   - Implemented a fade-to-black effect when the game is lost.
   - Lowered background volume

5. **Documentation:**
   - Added docstrings & comments to functions for better code documentation.
   - Added type annotations for variable clarity

icourt committed:
No need to add annotations to class attributes already passed via the init function.

Simpler annotations in GameController

icourt committed:
Minor Changes
- Private attributes used in GameController & Zombie to avoid accidental overwriting.
- Clean up annotations/imports
- Added docstring comments for clarification on Class/Module use cases
